pkg/endpoint/bpf.go

```diff
@@ -526,234 +526,238 @@ func (e *Endpoint) regenerateBPF(owner Owner, regenContext *regenerationContext)
 		}
-		datapathRegenCtxt.finalizeList.Append(finalizeFunc)
-		datapathRegenCtxt.revertStack.Push(revertFunc)
-	}
-	// At this point, traffic is no longer redirected to the proxy for
-	// now-obsolete redirects, since we synced the updated policy map above.
-	// It's now safe to remove the redirects from the proxy's configuration.
-	finalizeFunc, revertFunc = e.removeOldRedirects(owner, desiredRedirects, datapathRegenCtxt.proxyWaitGroup)
-	datapathRegenCtxt.finalizeList.Append(finalizeFunc)
-	datapathRegenCtxt.revertStack.Push(revertFunc)
-	stats.proxyConfiguration.End(true)
-
-	stats.prepareBuild.Start()
-
-	// Generate header file specific to this endpoint for use in compiling
-	// BPF programs for this endpoint.
-	if err = e.writeHeaderfile(nextDir, owner); err != nil {
-		stats.prepareBuild.End(false)
-		e.Unlock()
-		return 0, compilationExecuted, fmt.Errorf("unable to write header file: %s", err)
-	}
 
-	// Avoid BPF program compilation and installation if the headerfile for the endpoint
-	// or the node have not changed.
-	datapathRegenCtxt.bpfHeaderfilesHash, err = hashEndpointHeaderfiles(nextDir)
-	if err != nil {
-		e.getLogger().WithError(err).Warn("Unable to hash header file")
-		datapathRegenCtxt.bpfHeaderfilesHash = ""
-		datapathRegenCtxt.bpfHeaderfilesChanged = true
-	} else {
-		datapathRegenCtxt.bpfHeaderfilesChanged = (datapathRegenCtxt.bpfHeaderfilesHash != e.bpfHeaderfileHash)
-		e.getLogger().WithField(logfields.BPFHeaderfileHash, datapathRegenCtxt.bpfHeaderfilesHash).
-			Debugf("BPF header file hashed (was: %q)", e.bpfHeaderfileHash)
-	}
+		// Cache endpoint information so that we can release the endpoint lock.
+		if datapathRegenCtxt.bpfHeaderfilesChanged {
+			datapathRegenCtxt.epInfoCache = e.createEpInfoCache(nextDir)
+		} else {
+			datapathRegenCtxt.epInfoCache = e.createEpInfoCache(currentDir)
+		}
+		if datapathRegenCtxt.epInfoCache == nil {
+			stats.prepareBuild.End(false)
+			return fmt.Errorf("Unable to cache endpoint information")
+		}
 
-	// Cache endpoint information so that we can release the endpoint lock.
-	if datapathRegenCtxt.bpfHeaderfilesChanged {
-		datapathRegenCtxt.epInfoCache = e.createEpInfoCache(nextDir)
-	} else {
-		datapathRegenCtxt.epInfoCache = e.createEpInfoCache(currentDir)
-	}
-	if datapathRegenCtxt.epInfoCache == nil {
-		stats.prepareBuild.End(false)
-		e.Unlock()
-		err = fmt.Errorf("Unable to cache endpoint information")
+		// TODO: In Cilium v1.4 or later cycle, remove this.
+		os.RemoveAll(e.IPv6EgressMapPathLocked())
+		os.RemoveAll(e.IPv4EgressMapPathLocked())
+		os.RemoveAll(e.IPv6IngressMapPathLocked())
+		os.RemoveAll(e.IPv4IngressMapPathLocked())
+		return nil
+	}()
+
+	// Keep track of the side-effects of the regeneration that need to be
+	// reverted in case of failure.
+	// Also keep track of the regeneration finalization code that can't be
+	// reverted, and execute it in case of regeneration success.
+	defer func() {
```

Comment:
This depends on DryMode, doesn't it? at least, the old code would return when DryMode == true so we never made it to this defer. It also would be skipped in case of errors, so it might need to move below the error return below. In any case, I suspect it is more correct to call this in case of errors, but maybe not in dry mode.

Reply:
Fixed.